pkg/query/api/v1.go

```diff
@@ -288,27 +297,30 @@ func (api *API) query(r *http.Request) (interface{}, []error, *ApiError) {
 		warnmtx.Unlock()
 	}
 
+	raw := r.FormValue("query")
+	level.Debug(api.logger).Log("query", raw)
+
 	// We are starting promQL tracing span here, because we have no control over promQL code.
 	span, ctx := tracing.StartSpan(r.Context(), "promql_instant_query")
+	span.SetTag("query", raw)
 	defer span.Finish()
 
 	begin := api.now()
-	qry, err := api.queryEngine.NewInstantQuery(api.queryableCreate(enableDedup, 0, enablePartialResponse, warningReporter), r.FormValue("query"), ts)
+	qry, err := api.queryEngine.NewInstantQuery(api.queryableCreate(enableDedup, 0, enablePartialResponse, warningReporter), raw, ts)
 	if err != nil {
+		api.queryFailures.With(prometheus.Labels{"error": errorBadData}).Inc()
 		return nil, nil, &ApiError{errorBadData, err}
 	}
 
 	res := qry.Exec(ctx)
 	if res.Err != nil {
-		switch res.Err.(type) {
-		case promql.ErrQueryCanceled:
-			return nil, nil, &ApiError{errorCanceled, res.Err}
-		case promql.ErrQueryTimeout:
-			return nil, nil, &ApiError{errorTimeout, res.Err}
-		case promql.ErrStorage:
-			return nil, nil, &ApiError{ErrorInternal, res.Err}
-		}
-		return nil, nil, &ApiError{errorExec, res.Err}
+		// Cast Prometheus errors to Thanos error constants
+		e := toThanosError(res.Err)
+
+		level.Debug(api.logger).Log("query", raw, "error", res.Err.Error(), "duration", time.Since(begin).Seconds())
+		api.queryFailures.With(prometheus.Labels{"error": string(e)}).Inc()
+
+		return nil, nil, &ApiError{e, res.Err}
```

Comment:
This looks like wrong error handling to me as it's easy to make mistake and log/use metric on caller side, but I guess it's good enough for now.